Core SDK Implementation

dadheech115 opened the pull request:
This PR contains work for-
- Client Factory 
- HTTPClient
- Middleware Support
- Middlewares for Authentication and Redirect
- Network call mechanism with support for multiple request going through the pipeline simultaneously.
- And Unit Tests for all the implmentation

Commit:
Added client factory, http client and middleware support

Commit:
Changes project structure to indicate Core part

Commit:
Added Test target, Sample Test case for Session Manager class, Constants for error codes and response codes

Commit:
Added NSURLSession task wrappers to accommodate the cancellation of task as well

Commit:
Added Unit test cases for MSURLSessionManager class.

Commit:
Added MSURLSessionTask class's unit tests and a safe check while setting the next middleware

Commit:
Added Unit test for three different types of tasks - DataTask, DownloadTask and upload task.
Also removed unused code

Commit:
Added unit tests for MSHTTPClient methods and added some 'nonnull' attributes for method arguments in this class

Commit:
Added MSClientFactory unit tests and corrected a method name

Commit:
Added Unit test cases for MSURLSessionTaskDelegate class and removed unused code from there

Commit:
Added mock unit test cases of workloads to cover the pipe line of execution and reorganised the test cases directory structure

Commit:
Added Unit tests to cover remaining completion blocks in MSURLSessionTask and MSURLSessionManager

Commit:
Added middleware factory to help in creation of middleware instances, removed unwanted logs and added consistency in the way method names are written

Commit:
Handled the error case in Auth Middleware and Handled bool checking gracefully in test cases by moving the bool setting and checking mechanism to parent class

Commit:
Improved code readability by making use of curly braces and comment on top of each file consistent.
Restructured the code of MSURLSessionManager class's execute function to remove duplicity of code.

Commit:
Added Redirect Handler as part of middleware chain

Commit:
Changed framework's scheme to shared. Changed authentication provider's method signature and changes some of the workload tests

Commit:
Removed OCMock pod dependancy and added as static lib

Commit:
Incorporating review comments -
# Copy attribute for completionHandlers
# changed setNext implementation
# code consistency in terms of curly braces, whitespaces, line breaks, whitespace after comma
# nil check on stored completionHandlers

Commit:
Added unit tests for redirect handler and authentication middleware

Commit:
Renamed MSAuthenticationMiddleware to MSAuthenticationHandler

Commit:
Added sdk version header and unit tests

Commit:
Added README file

Commit:
Changed project version, updated xcode compatibility and min os support

Commit:
Changed request to hold mutable copy and added unit tests to test the same

deepak2016 commented on MSGraphCoreSDK/MSGraphCoreSDK/Common/MSConstants.m:
```diff
@@ -0,0 +1,16 @@
+#import "MSConstants.h"
+
+NSString *const MSGraphBaseURL = @"https://graph.microsoft.com/v1.0";
+NSString *const MSHeaderSdkVersion = @"SdkVersion";
```
Should sdk version be "SdkVersion-0.1.0"?

deepak2016 commented on MSGraphCoreSDK/MSGraphCoreSDK/Authentication/MSAuthenticationProvider.h:
```diff
@@ -0,0 +1,23 @@
+@protocol MSAuthenticationProvider <NSObject>
+
+/**
+ Gets the access token. This method should be implmented by a class which should provide the capability of providing the access token.
```
minor: spelling correction in the comment  for - "implemented"

dadheech115 replied:
Sure. Will correct that.

deepak2016 commented on MSGraphCoreSDK/MSGraphCoreSDK/Middleware/Implementations/HTTPProvider/MSURLSessionManager.m:
```diff
@@ -0,0 +1,253 @@
+
+- (void)setNext:(id<MSGraphMiddleware>)nextMiddleware
+{
+    if(_nextMiddleware)
```
As this middleware is always going to be the last middleware, don't provide mechanism to set next middleware.
Remove this implementation?

dadheech115 replied:
a stub will suffice in that case I think?
Can't remove the implementation as the method is part of MSGraphMiddleware protocol and this class conforms to it.

deepak2016 replied:
Yes, stub should be fine. Maybe with some assert warning that setnext is not allowed in this middleware.

Commit:
Resolved review comments
# Spelling mistake in MSAuthenticationProvider
# Added stub for MSGraphMiddleware's setNext method in MSURLSessionManager and Unit test for the same
